### PR TITLE
Support http2.

### PR DESCRIPTION
**Pull request by sogaani**
Hi.

This PR enable http/2, if EXPOSE_HTTP2 env is true.

This PR still not support ALPN(as @kornelski mentioned in https://github.com/visionmedia/superagent/issues/980).
I'd prefer to add ALPN functionality after merging this PR and [express http/2 PR](https://github.com/expressjs/express/pull/3390).

### CODE REVIEW

**Comment by kornelski** [comment]
Thank you! That's a very useful functionality.

**Comment by kornelski** [comment]
Does `EXPOSE_HTTP2` affect ability to make HTTP/1 requests? Is Node going to fall back automatically? I wonder if we need to give more runtime control over this.

File: lib/node/http2wrapper.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  const sessionOptions = Object.assign({}, options);
+  if (options.socketPath) {
+    sessionOptions.socketPath = options.socketPath;
+    sessionOptions.createConnection = this.createUnixConnection.bind(this);
+  }
+
+  this._headers = {};
+
+  const session = http2.connect(`${protocol}://${host}:${port}`, sessionOptions);
+  this.setHeader('host', `${host}:${port}`)
```
Comment (kornelski) [inline lib/node/http2wrapper.js]: Is it necessary here? In normal implementation we allow `options.host` to be different than host header, because it's useful to do `request.get('http://[ip address]').host('example.com')` to force request to connect to a particular IP address.

File: lib/node/http2wrapper.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+ * Inherit from `Stream` (which inherits from `EventEmitter`).
+ */
+util.inherits(Request, Stream);
+
+Request.prototype.createUnixConnection = function (authority, options) {
+  switch (this.protocol) {
+    case 'http':
+      return net.connect(options.socketPath);
+    case 'https':
+      options.ALPNProtocols = ['h2'];
```
Comment (kornelski) [inline lib/node/http2wrapper.js]: AFAIK browsers also send http/1 here. Do we support a fallback?

File: lib/node/http2wrapper.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  const keys = Object.keys(headers);
+  const http2Headers = {};
+  for (var i = 0; i < keys.length; i++) {
+    let key = keys[i];
+    let value = headers[key];
+    key = key.toLowerCase();
+    switch (key) {
+      case HTTP2_HEADER_HOST:
+        key = HTTP2_HEADER_AUTHORITY;
+        value = value.startsWith('http') ? parse(value).host : value;
```
Comment (kornelski) [inline lib/node/http2wrapper.js]: Why is such conditional parsing needed?

File: test/node/pipe.js [hunk 4]
```diff
@@ -12,13 +17,19 @@ app.get("/", (req, res) => {
 });
 
 app.post("/", (req, res) => {
-  res.send(req.body);
+  if (process.env.EXPOSE_HTTP2){
+    res.set('content-type', 'application/json');
```
Comment (kornelski) [inline test/node/pipe.js]: Why is this needed?

File: test/node/unix-sockets.js [hunk 5]
```diff
@@ -94,6 +105,7 @@ describe("[unix-sockets] https", () => {
         .get(`${base}/request/path`)
         .ca(cert)
         .end((err, res) => {
+          console.log(err);
```
Comment (kornelski) [inline test/node/unix-sockets.js]: 👀

File: test/node/user-agent.js [hunk 6]
```diff
@@ -6,7 +6,7 @@ const base = setup.uri;
 
 describe("req.get()", () => {
   it("should set a default user-agent", () =>
-    request.get(`${base}/ua`).then(res => {
+    request.put(`${base}/ua`).then(res => {
```
Comment (kornelski) [inline test/node/user-agent.js]: Can you explain this change?

File: test/support/server.js [hunk 7]
```diff
@@ -17,7 +42,15 @@ app.all('/url', (req, res) => {
 });
 
 app.all('/echo', (req, res) => {
-  res.writeHead(200, req.headers);
+  let headers = req.headers;
+  if(process.env.EXPOSE_HTTP2){
+    Object.keys(headers).forEach(function(name) {
+      if(isPseudoHeader(name)){
+        delete headers[name];
```
Comment (kornelski) [inline test/support/server.js]: Maybe this should translate the header to HTTP/1 header, or allow both in the test using it?

File: lib/node/http2wrapper.js (hunk 2 above)
Comment (sogaani) [reply lib/node/http2wrapper.js]: Current implement does not support fall back. I can implement fallback, I plan to take fallback after this PR. Or Should I integrate  with this PR?

**Comment by kornelski** [comment]
@focusaurus do you have a preferences how HTTP/1 vs 2 API should look like?

File: lib/node/http2wrapper.js (hunk 3 above)
Comment (sogaani) [reply lib/node/http2wrapper.js]: This is add for compatibility with Http/1. HTTP2_HEADER_AUTHORITY(alternative Host in http/2) only support domain string, otherwise it raise error. However, I met Host header with schema(like Host: http://example.com) in some express tests.

File: test/node/pipe.js (hunk 4 above)
Comment (sogaani) [reply test/node/pipe.js]: http/2 module have not full compatibility with http/1 module. http/2 imcommingHttpRequest does not have body property.

File: test/node/unix-sockets.js (hunk 5 above)
Comment (sogaani) [reply test/node/unix-sockets.js]: My misstake! I will fix 👍

File: test/node/user-agent.js (hunk 6 above)
Comment (sogaani) [reply test/node/user-agent.js]: http/2 client module cannot send body with head or get Method. But, it seems to not need here. I check this again later.

File: test/support/server.js (hunk 7 above)
Comment (sogaani) [reply test/support/server.js]: HTTP/2 use some headers to control protocol(e.g. :method, :authority, :path ...) and it can not set by user. I just delete those headers here.

File: lib/node/http2wrapper.js (hunk 1 above)
Comment (sogaani) [reply lib/node/http2wrapper.js]: I think current implementation supports `request.get('http://[ip address]').host('example.com')`. I will check later 👍 .

**Comment by focusaurus** [comment]
Hmm. Am I interpreting the discussion correctly? So if a user sets `EXPOSE_HTTP2=true` they can only connect to http2 servers from that entire node process and connections to http/1 servers will fail? That sounds untenable. I think I'd prefer leaving HTTP/1 as the default and exposing an API to use http/2 on a per-request basis. For that I think the `.http2()` method name is a good API.

File: lib/node/http2wrapper.js (hunk 3 above)
Comment (kornelski) [reply lib/node/http2wrapper.js]: So it would be better to improve the check, otherwise host like `httparchive.org` would be misparsed :)

```js
/^https?:\/\//.test(value)
```

**Comment by sogaani** [comment]
@focusaurus Thank you for feedback!.
>  I think I'd prefer leaving HTTP/1 as the default and exposing an API to use http/2 on a per-request basis. For that I think the .http2() method name is a good API.

It is reasonable. I change this soon. 👍

File: test/node/user-agent.js (hunk 6 above)
Comment (sogaani) [reply test/node/user-agent.js]: I test locally and noticed that this change is not needed. I will revert this change.

File: lib/node/http2wrapper.js (hunk 1 above)
Comment (sogaani) [reply lib/node/http2wrapper.js]: Still not send header here and we can change `Host` header by `set` method. I add test for this.

File: lib/node/http2wrapper.js (hunk 2 above)
Comment (kornelski) [reply lib/node/http2wrapper.js]: I see. If there's no fallback, then that's OK.

**Comment by sogaani** [comment]
I addressed review and change the API to exposing an HTTP/2 API.

Superagent can be enabled http2 by http2 field:

```js
    const request = require('superagent');
    request.http2 =true;
    request
      .get('http://example.com/search')
      .then(res => {

      });
```

A request can be enabled http2 by http2 method:

```js
    const request = require('superagent');
    request
      .get('http://example.com/search')
      .http2()
      .then(res => {

      });
```

**Comment by sogaani** [comment]
ping @kornelski  @focusaurus

**Comment by kornelski** [comment]
Thanks

**Comment by focusaurus** [comment]
Awesome contribution @sogaani !
